modules/message-lane/src/lib.rs

```diff
@@ -66,6 +66,11 @@ pub trait Trait<I = DefaultInstance>: frame_system::Trait {
 	/// confirmed. The reason is that if you want to use lane, you should be ready to pay
 	/// for it.
 	type MaxMessagesToPruneAtOnce: Get<MessageNonce>;
+	/// Maximal number of messages in the 'unconfirmed' state at inbound lane. Unconfirmed
+	/// message at inbound lane is the message that has been: sent, delivered and dispatched.
+	/// Its delivery confirmation is still pending. This limit is introduced to bound maximal
+	/// number of relayers-ids in the inbound lane state.
+	type MaxUnconfirmedMessagesAtInboundLane: Get<MessageNonce>;
```

Comment:
While this works I really doubt it's the right abstraction. It not only limits the number of "unconfirmed messages", but it actually also limits the number of messages you can dispatch in a single transaction (even though it's a single relayer).

A really simple optimisation would be to store `VecDeque<(StartNonce, Option<EndNonce>, Relayer)>` instead of duplicating the relayer for each nonce it delivered. We really expect the relayers to dispatch multiple messages at once, so IMHO it makes sense to optimize that. Thanks to this, if a relayer delivers a batch, it will only occupy one slot in that collection, so the `MaxUncofirmedMessagesAtInboundeLane` wouldn't restrict the message delivery that much.

Reply:
> While this works I really doubt it's the right abstraction. It not only limits the number of "unconfirmed messages", but it actually also limits the number of messages you can dispatch in a single transaction (even though it's a single relayer).

If you care about possible number of mesages in a single transaction, make this at least `MaxBlockWeight / MaxExtrinsicWeight`. But in practice it should be much larger, or we will be blocked by GRANDPA authorities + headers relays.

> A really simple optimisation would be to store VecDeque<(StartNonce, Option<EndNonce>, Relayer)> instead of duplicating the relayer for each nonce it delivered. We really expect the relayers to dispatch multiple messages at once, so IMHO it makes sense to optimize that. Thanks to this, if a relayer delivers a batch, it will only occupy one slot in that collection, so the MaxUncofirmedMessagesAtInboundeLane wouldn't restrict the message delivery that much.

Yes - this may be done, but since it is only a potential optimization - let's file a separate issue for that.

Reply:
> MaxBlockWeight / MaxExtrinsicWeight

This is obviously wrong, sorry. Should be something like `MaxBlockWeight / Avg(MaxExtrinsicWeight)`. Nevertheless, it is still relevant when talking about messages/tx.

Reply:
Extracted as #406